x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java

```diff
@@ -305,17 +333,36 @@ static class StagnatingIndicesFinder {
         public List<IndexMetadata> find() {
             var metadata = clusterService.state().metadata();
             var now = nowSupplier.getAsLong();
+            var clusterSettings = clusterService.getClusterSettings();
+            var defaultMaxTimeOnAction = clusterSettings.get(MAX_TIME_ON_ACTION_SETTING);
+            var defaultMaxTimeOnStep = clusterSettings.get(MAX_TIME_ON_STEP_SETTING);
+            var defaultMaxRetriesPerStep = clusterSettings.get(MAX_RETRIES_PER_STEP_SETTING);
+
             return metadata.indices()
                 .values()
                 .stream()
                 .filter(metadata::isIndexManagedByILM)
-                .filter(md -> isStagnated(rules, now, md))
+                .filter(md -> isStagnated(rules, defaultMaxTimeOnAction, defaultMaxTimeOnStep, defaultMaxRetriesPerStep, now, md))
                 .toList();
         }
     }
 
-    static boolean isStagnated(Collection<RuleConfig> rules, Long now, IndexMetadata indexMetadata) {
-        return rules.stream().anyMatch(r -> r.test(now, indexMetadata));
+    static boolean isStagnated(
+        Collection<RuleCreator> rules,
+        TimeValue defaultMaxTimeOnAction,
+        TimeValue defaultMaxTimeOnStep,
+        Long defaultMaxRetriesPerStep,
+        Long now,
+        IndexMetadata indexMetadata
+    ) {
+        return rules.stream()
+            .map(rc -> rc.create(defaultMaxTimeOnAction, defaultMaxTimeOnStep, defaultMaxRetriesPerStep))
+            .anyMatch(r -> r.test(now, indexMetadata));
+    }
+
+    @FunctionalInterface
+    interface RuleCreator {
+        RuleConfig create(TimeValue defaultMaxTimeOnAction, TimeValue defaultMaxTimeOnStep, Long defaultMaxRetriesPerStep);
```

Comment:
Do we need this interface mostly because the rules are statically defined? 

The current design will keep recreating the rules for every index, which is sub-optimal I think? (i.e. if we have more than 50_000 indices, even if all healthy and not stagnated)

What if we made the rules an instance variable of `IlmHealthIndicatorService` ? ( we only create one instance of `IlmHealthIndicatorService` anyway) We could probably access `IlmHealthIndicatorService` instance variables in this case and can potentially follow the pattern where we react to cluster settings updates and only then recreate the rules (in constructor we can read the settings and create the rules and then add something like this): 

```
clusterService.getClusterSettings()
            .addSettingsUpdateConsumer(MAX_TIME_ON_ACTION_SETTING, this::recreateTheRules);
```

Reply:
good idea, will implement that!

Reply:
to be honest, I added the `RuleCreator` interface to avoid having  `TriFunction<TimeValue, TimeValue, Long, RuleConfig>` all over the place... anyway, I did the change that subscribes to settings updates.